pkg/cleanup/controller/controller.go

```diff
@@ -0,0 +1,262 @@
+package controller
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	coreinformers "k8s.io/client-go/informers/core/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/kubernetes/scheme"
+	corev1 "k8s.io/client-go/kubernetes/typed/core/v1"
+	corelisters "k8s.io/client-go/listers/core/v1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/tools/record"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog/v2"
+
+	"sigs.k8s.io/sig-storage-local-static-provisioner/pkg/common"
+)
+
+type CleanupController struct {
+	client kubernetes.Interface
+
+	// pvQueue is a rate-limited delayed queue. PVs names are added
+	// to this queue when its corresponding Node is deleted. The queue's delay allows to wait
+	// for the Node to come back up before cleaning up resources.
+	pvQueue workqueue.RateLimitingInterface
+
+	nodeLister       corelisters.NodeLister
+	nodeListerSynced cache.InformerSynced
+
+	pvLister       corelisters.PersistentVolumeLister
+	pvListerSynced cache.InformerSynced
+
+	eventRecorder record.EventRecorder
+	broadcaster   record.EventBroadcaster
+
+	// storageClassName is the name of the StorageClass that PVs and PVCs
+	// must belong to in order to be cleaned up.
+	storageClassName string
+
+	// delay is the amount of time to wait after Node deletion to cleanup resources.
+	delay time.Duration
+
+	// stalePVDiscoveryInterval is how often to scan for and delete PVs with affinity to a deleted Node.
+	stalePVDiscoveryInterval time.Duration
+}
+
+func NewCleanupController(client kubernetes.Interface, pvInformer coreinformers.PersistentVolumeInformer, nodeInformer coreinformers.NodeInformer, storageClassName string, delay time.Duration, stalePVDiscoveryInterval time.Duration) *CleanupController {
+	broadcaster := record.NewBroadcaster()
+	eventRecorder := broadcaster.NewRecorder(scheme.Scheme, v1.EventSource{Component: fmt.Sprintf("cleanup-controller")})
+
+	controller := &CleanupController{
+		client:           client,
+		storageClassName: storageClassName,
+		// Delayed queue with rate limiting
+		pvQueue: workqueue.NewRateLimitingQueueWithConfig(
+			workqueue.DefaultControllerRateLimiter(),
+			workqueue.RateLimitingQueueConfig{
+				Name: "stale",
```

Comment:
Be more specific in naming this to describe what it does. The name of the workqueue shows up in metrics.

Reply:
Changed to stalePVQueue.